Mathlib/MeasureTheory/Measure/EverywherePos.lean

```diff
@@ -0,0 +1,296 @@
+    true_and, (hs.diff hs.everywherePosSubset).measure_eq_iSup_isCompact_of_ne_top A,
+    ENNReal.iSup_eq_zero]
+  intro k hk h'k
+  exact measure_eq_zero_of_subset_diff_everywherePosSubset h'k hk
+
+/-- In a space with an inner regular measure, the everywhere positive subset of a measurable set
+is itself everywhere positive. This is not obvious as `μ.everywherePosSubset s` is defined as
+the points whose neighborhoods intersect `s` along positive measure subsets, but this does not
+say they also intersect `μ.everywherePosSubset s` along positive measure subsets. -/
+lemma isEverywherePos_everywherePosSubset
+    [OpensMeasurableSpace α] [InnerRegular μ] (hs : MeasurableSet s) :
+    μ.IsEverywherePos (μ.everywherePosSubset s) := by
+  intro x hx n hn
+  rcases mem_nhdsWithin_iff_exists_mem_nhds_inter.1 hn with ⟨u, u_mem, hu⟩
+  have A : 0 < μ (u ∩ s) := by
+    have : u ∩ s ∈ 𝓝[s] x := by rw [inter_comm]; exact inter_mem_nhdsWithin s u_mem
+    exact hx.2 _ this
+  have B : (u ∩ μ.everywherePosSubset s : Set α) =ᵐ[μ] (u ∩ s : Set α) :=
+    ae_eq_set_inter (ae_eq_refl _) (everywherePosSubset_ae_eq hs)
+  rw [← B.measure_eq] at A
+  exact A.trans_le (measure_mono hu)
+
+/-- In a space with an inner regular measure for finite measure sets, the everywhere positive subset
+of a measurable set of finite measure is itself everywhere positive. This is not obvious as
+`μ.everywherePosSubset s` is defined as the points whose neighborhoods intersect `s` along positive
+measure subsets, but this does not say they also intersect `μ.everywherePosSubset s` along positive
+measure subsets.-/
+lemma isEverywherePos_everywherePosSubset_of_measure_ne_top
+    [OpensMeasurableSpace α] [InnerRegularCompactLTTop μ] (hs : MeasurableSet s) (h's : μ s ≠ ∞) :
+    μ.IsEverywherePos (μ.everywherePosSubset s) := by
+  intro x hx n hn
+  rcases mem_nhdsWithin_iff_exists_mem_nhds_inter.1 hn with ⟨u, u_mem, hu⟩
+  have A : 0 < μ (u ∩ s) := by
+    have : u ∩ s ∈ 𝓝[s] x := by rw [inter_comm]; exact inter_mem_nhdsWithin s u_mem
+    exact hx.2 _ this
+  have B : (u ∩ μ.everywherePosSubset s : Set α) =ᵐ[μ] (u ∩ s : Set α) :=
+    ae_eq_set_inter (ae_eq_refl _) (everywherePosSubset_ae_eq_of_measure_ne_top hs h's)
+  rw [← B.measure_eq] at A
+  exact A.trans_le (measure_mono hu)
+
+lemma IsEverywherePos.smul_measure (hs : IsEverywherePos μ s) {c : ℝ≥0∞} (hc : c ≠ 0) :
+    IsEverywherePos (c • μ) s :=
+  fun x hx n hn ↦ by simpa [hc.bot_lt, hs x hx n hn] using hc.bot_lt
+
+lemma IsEverywherePos.smul_measure_nnreal (hs : IsEverywherePos μ s) {c : ℝ≥0} (hc : c ≠ 0) :
+    IsEverywherePos (c • μ) s :=
+  hs.smul_measure (by simpa using hc)
+
+/-- If two measures coincide locally, then a set which is everywhere positive for the former is
+also everywhere positive for the latter. -/
+lemma IsEverywherePos.of_forall_exists_nhds_eq (hs : IsEverywherePos μ s)
+    (h : ∀ x ∈ s, ∃ t ∈ 𝓝 x, ∀ u ⊆ t, ν u = μ u) : IsEverywherePos ν s := by
```

Comment:
Does this concept of two measures coinciding locally occur enough that we should have a definition + API for it?

Reply:
I don't think so, because we have just one application for now. I will reconsider when we have a second application :-)

Reply:
Measures form a sheaf, and this is an important property, that allows to construct measures locally. (In fact, it's even a sheaf for a Grothendieck topology given by coverings by locally closed subsets — the definition of a covering is standard, but stronger than just the union is everything).

Reply:
It really depends on the sigma-algebra you consider. On the sigma algebra of Baire sets (generated by compact G-delta sets), measures form indeed a sheaf. On the Borel sigma-algebra, they don't (an example I like is: let \mu be zero on sets which are contained in a countable union of balls of radius 1, and +\infty otherwise. On a non-separable Hilbert space, this is a measure which is locally zero, with trivial support, but still nonzero).